src/storage/store/engine.rs

```diff
@@ -212,11 +241,16 @@ impl ShardEngine {
                 break;
             }
 
-            while messages.len() < self.max_messages_per_block as usize {
-                match self.messages_rx.try_recv() {
-                    Ok(msg) => messages.push(msg),
-                    Err(mpsc::error::TryRecvError::Empty) => break,
-                    Err(err) => return Err(EngineError::from(err)),
+            let capacity = self.max_messages_per_block as usize - messages.len();
+
+            if capacity > 0 {
+                let mempool = self.mempool.read().await;
```

Comment:
Our philosophy has been to avoid locks and prefer channels for cross thread communication.